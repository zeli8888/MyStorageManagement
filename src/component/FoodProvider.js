import React, { Component, useState, useEffect } from 'react';
import IngredientService from '../service/IngredientService'
import DishService from '../service/DishService'

export const FoodContext = React.createContext();  //exporting context object

const FoodProvider = ({ children }) => {
    const [ingredients, setIngredients] = useState([]);
    const [dishes, setDishes] = useState([]);

    useEffect(() => {
        IngredientService.getAllIngredients().then(response => { setIngredients(response.data) })
        DishService.getAllDishes().then(response => { setDishes(response.data) })
    }, []);

    return (
        <FoodContext.Provider value={
            {
                ingredients: ingredients,
                setIngredients: setIngredients,
                dishes: dishes,
                setDishes: setDishes
            }}>
            {children
                //this indicates that all the child tags with MyProvider as Parent can access the global store  
            }
        </FoodContext.Provider>)
}

export default FoodProvider;