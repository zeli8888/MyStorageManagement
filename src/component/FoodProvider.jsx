import React, { useState, useEffect } from 'react';
import ingredientService from '../service/IngredientService'
import dishService from '../service/DishService'

export const FoodContext = React.createContext();  //exporting context object

const FoodProvider = ({ children }) => {
    const [allIngredients, setAllIngredients] = useState([]);
    const [allDishes, setAllDishes] = useState([]);

    useEffect(() => {
        ingredientService.getAllIngredients().then(response => { setAllIngredients(response.data) }).catch(error => { console.log(error) });
        dishService.getAllDishes().then(response => { setAllDishes(response.data) }).catch(error => { console.log(error) });
    }, []);

    return (
        <FoodContext.Provider value={
            {
                allIngredients: allIngredients,
                setAllIngredients: setAllIngredients,
                allDishes: allDishes,
                setAllDishes: setAllDishes
            }}>
            {children
                //this indicates that all the child tags with MyProvider as Parent can access the global store  
            }
        </FoodContext.Provider>)
}

export default FoodProvider;