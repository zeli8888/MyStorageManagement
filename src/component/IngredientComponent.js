import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { DataGrid } from '@mui/x-data-grid';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import IngredientService from '../service/IngredientService'
import { FoodContext } from './FoodProvider';
import { DeletionConfirmationComponent, SearchComponent } from './MyComponents';
import Alert from '@mui/material/Alert';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
const IngredientComponent = function () {
    const navigate = useNavigate();
    const [ingredients, setIngredients] = useState([]);
    const { setAllIngredients } = React.useContext(FoodContext);
    const [ingredientUpdating, setIngredientUpdating] = useState();
    const [addingIngredient, setAddingIngredient] = useState(false);
    const [ingredientToDelete, setIngredientToDelete] = useState();
    const [ingredientAlert, setIngredientAlert] = useState();

    useEffect(() => {
        refreshIngredients();
    }, [navigate]);

    const refreshIngredients = () => {
        IngredientService.getAllIngredients().then(response => {
            setIngredients(response.data);
            setAllIngredients(response.data);
        }).catch(error => {
            console.log(error);
        });
    }

    const searchIngredients = (searchText) => {
        IngredientService.searchIngredients(searchText).then(response => {
            setIngredients(response.data);
        }).catch(error => {
            console.log(error);
        });
    }

    const updateIngredient = (data) => {
        IngredientService.updateIngredient(data.ingredientId, data).then(response => {
            setIngredientUpdating();
            setAddingIngredient(false);
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredient " + data.ingredientName + " updated successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const addIngredient = (data) => {
        if (ingredientUpdating) return updateIngredient(data);
        IngredientService.addIngredient(data).then(response => {
            setAddingIngredient(false);
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredient " + data.ingredientName + " added successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const deleteIngredient = (ingredientId) => {
        IngredientService.deleteIngredient(ingredientId).then(response => {
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredient " + ingredientToDelete.ingredientName + " deleted successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const columns = [
        {
            field: 'ingredientId',
            headerName: 'Id',
            with: 0,
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientName',
            headerName: 'Ingredient',
            flex: 2,
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientStorage',
            headerName: 'Storage',
            flex: 2,
            type: 'number',
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientCost',
            headerName: 'Cost',
            flex: 2,
            type: 'number',
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientDesc',
            headerName: 'Description',
            flex: 3,
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'update',
            headerName: 'Update',
            headerAlign: 'center',
            flex: 3,
            renderCell: (params) => (
                <Button variant="contained" color="success" onClick={
                    () => {
                        setIngredientUpdating(params.row);
                        setAddingIngredient(true);
                    }
                } > Update</Button >
            )
        },
        {
            field: 'delete',
            headerName: 'Delete',
            headerAlign: 'center',
            flex: 3,
            renderCell: (params) => (
                <Button variant="contained" color="error" onClick={
                    () => setIngredientToDelete(params.row)
                }>Delete</Button>
            )
        }
    ];

    return (
        <Box sx={{ flexGrow: 1 }}>
            <Grid container spacing={2} >
                <Grid size={12}>
                    {ingredientAlert && <Alert severity={ingredientAlert.severity} onClose={() => { setIngredientAlert(null) }}>{ingredientAlert.message}</Alert>}
                </Grid>
                <Grid size={12}>
                    <SearchComponent onSearch={searchIngredients} />
                </Grid>
                <Grid size={12}>
                    <DataGrid
                        autoWidth
                        rows={ingredients}
                        columns={columns}
                        initialState={{
                            sorting: {
                                sortModel: [{
                                    field: 'ingredientStorage',
                                    sort: 'asc',
                                }],
                            },
                            columns: {
                                columnVisibilityModel: {
                                    ingredientId: false,
                                },
                            },
                            pagination: {
                                paginationModel: {
                                    pageSize: 10,
                                },
                            },
                        }}
                        pageSizeOptions={[10, 20]}
                        checkboxSelection
                        getRowId={(row) => row.ingredientId}
                        sx={{
                            boxShadow: 2,
                            border: 2,
                            borderColor: 'lightgrey',
                            width: '100%',
                            '& .MuiDataGrid-cell': {
                                textAlign: 'center', // Center the text in each cell
                            }
                        }}
                    />
                </Grid>
                <Button variant="contained" color="success"
                    onClick={() => {
                        setIngredientUpdating();
                        setAddingIngredient(true);
                    }}
                    sx={{
                        marginRight: 0,
                        marginLeft: 'auto'
                    }}
                >
                    New Ingredient
                </Button>
            </Grid>
            <Dialog
                open={addingIngredient}
                onClose={() => {
                    setIngredientUpdating();
                    setAddingIngredient(false);
                }}
                maxWidth="xs"
                fullWidth={true}
                slotProps={{
                    paper: {
                        component: 'form',
                        onSubmit: (event) => {
                            event.preventDefault();
                            const formData = new FormData(event.currentTarget);
                            const formJson = Object.fromEntries(formData.entries());
                            addIngredient(formJson);
                        },
                    },
                }}
            >
                <DialogTitle>{ingredientUpdating ? "Update Ingredient" : "New Ingredient"}</DialogTitle>
                <DialogContent>
                    {ingredientUpdating && <input type="hidden" name="ingredientId" value={ingredientUpdating.ingredientId} />}
                    <TextField
                        autoFocus
                        required
                        margin="dense"
                        name="ingredientName"
                        label="Ingredient Name"
                        type="text"
                        fullWidth
                        variant="standard"
                        defaultValue={ingredientUpdating ? ingredientUpdating.ingredientName : ""}
                    />
                    <TextField
                        required
                        margin="dense"
                        name="ingredientStorage"
                        label="Ingredient Storage"
                        type="number"
                        fullWidth
                        variant="standard"
                        defaultValue={ingredientUpdating ? ingredientUpdating.ingredientStorage : ""}
                    />
                    <TextField
                        margin="dense"
                        name="ingredientCost"
                        label="Ingredient Cost"
                        type="number"
                        fullWidth
                        variant="standard"
                        defaultValue={ingredientUpdating ? ingredientUpdating.ingredientCost : ""}
                    />
                    <TextField
                        margin="dense"
                        name="ingredientDesc"
                        label="Ingredient Description"
                        type="text"
                        fullWidth
                        variant="standard"
                        defaultValue={ingredientUpdating ? ingredientUpdating.ingredientDesc : ""}
                    />
                </DialogContent>
                <DialogActions>
                    <Button onClick={() => {
                        setAddingIngredient(false);
                        setAddingIngredient(false);
                    }}>Cancel</Button>
                    <Button type="submit">Save</Button>
                </DialogActions>
            </Dialog>
            <DeletionConfirmationComponent warningMessage={ingredientToDelete ? "Are you sure you want to delete ingredient " + ingredientToDelete.ingredientName + "?" : "Processing"}
                open={ingredientToDelete} onClose={() => setIngredientToDelete(null)}
                onConfirm={() => {
                    deleteIngredient(ingredientToDelete.ingredientId);
                    setIngredientToDelete(null);
                }} />
        </Box>
    )
}

export default IngredientComponent;