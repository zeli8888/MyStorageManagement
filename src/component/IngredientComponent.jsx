import { useContext, useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { DataGrid } from '@mui/x-data-grid';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import ingredientService from '../service/IngredientService'
import { FoodContext } from './FoodProvider';
import { DeletionConfirmationComponent, EnhancedTableToolbar, NumberInput } from './utils';
import Alert from '@mui/material/Alert';
import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import moment from 'moment';
import InputLabel from '@mui/material/InputLabel';
import dishRecordService from '../service/DishRecordService'

const ModalContentIngredients = function (props) {
    const { ingredientUpdating } = props;
    return (
        <div>
            <DialogTitle>{ingredientUpdating ? "Update Ingredient" : "New Ingredient"}</DialogTitle>
            <DialogContent>
                {ingredientUpdating && <input type="hidden" name="ingredientId" value={ingredientUpdating.ingredientId} />}
                <TextField
                    autoFocus
                    required
                    margin="dense"
                    name="ingredientName"
                    label="Ingredient Name"
                    type="text"
                    fullWidth
                    variant="standard"
                    defaultValue={ingredientUpdating ? ingredientUpdating.ingredientName : ""}
                />
                <NumberInput
                    required
                    margin="dense"
                    name="ingredientStorage"
                    label="Ingredient Storage"
                    // NumericFormat props 
                    thousandSeparator
                    decimalScale={2}
                    fixedDecimalScale
                    //
                    fullWidth
                    variant="standard"
                    defaultValue={ingredientUpdating ? ingredientUpdating.ingredientStorage : ""}
                />
                <NumberInput
                    margin="dense"
                    name="ingredientCost"
                    label="Ingredient Cost"
                    // NumericFormat props 
                    thousandSeparator
                    decimalScale={2}
                    fixedDecimalScale
                    //
                    fullWidth
                    variant="standard"
                    defaultValue={ingredientUpdating ? ingredientUpdating.ingredientCost : ""}
                />
                <TextField
                    margin="dense"
                    name="ingredientDesc"
                    label="Ingredient Description"
                    type="text"
                    fullWidth
                    variant="standard"
                    defaultValue={ingredientUpdating ? ingredientUpdating.ingredientDesc : ""}
                />
            </DialogContent>
        </div>
    )
}

const ModalContentRecord = function (props) {
    const { ingredientIdsForRecord, ingredients } = props;
    const ingredientNamesForRecord = ingredientIdsForRecord.map(
        ingredientId => ingredients.find(ingredient => ingredient.ingredientId == ingredientId).ingredientName
    );
    return (
        <div>
            <DialogTitle>New Record</DialogTitle>
            <DialogContent>
                <TextField
                    autoFocus={false}
                    margin="dense"
                    name={"dishRecordDesc"}
                    label={"Record Description"}
                    type="text"
                    fullWidth
                    variant="standard"
                    defaultValue={ingredientNamesForRecord.join(", ")}
                />
                <TextField
                    margin="dense"
                    name="dishRecordTime"
                    label="Record Time"
                    type="datetime-local"
                    fullWidth
                    variant="standard"
                    defaultValue={moment(new Date()).format('YYYY-MM-DD HH:mm')}
                />
                <Box mt={4} />
                <InputLabel id="add-ingredient-for-record">{"Ingredients Used"}</InputLabel>
                <Grid container spacing={2} >
                    {ingredientIdsForRecord.map((ingredientId, index) => {
                        const ingredientName = ingredientNamesForRecord[index];
                        return (
                            <div key={'ingredient-' + ingredientId}>
                                <input type="hidden" name={'ingredientName-' + ingredientId} value={ingredientName} />
                                <NumberInput
                                    name={'ingredientAmount-' + ingredientId}
                                    label={ingredientName + " Amount"}
                                    variant="standard"
                                    color="success"
                                    margin="dense"
                                    // NumericFormat props 
                                    thousandSeparator
                                    decimalScale={2}
                                    fixedDecimalScale
                                    defaultValue={1}
                                    required
                                />
                            </div>
                        );
                    })}
                </Grid>
            </DialogContent>
        </div>
    )
}

const IngredientsModal = function (props) {
    const { open, addingDishRecord, ingredientUpdating, ingredientIdsForRecord, ingredients, onClose, onSubmit } = props;
    return (
        <Dialog
            open={open}
            onClose={onClose}
            maxWidth="xs"
            fullWidth={true}
            slotProps={{
                paper: {
                    component: 'form',
                    onSubmit: onSubmit,
                },
            }}
        >
            {!addingDishRecord ?
                <ModalContentIngredients ingredientUpdating={ingredientUpdating} /> :
                <ModalContentRecord ingredientIdsForRecord={ingredientIdsForRecord} ingredients={ingredients} />
            }

            <DialogActions>
                <Button onClick={onClose} data-testid="modal-cancel-button">Cancel</Button>
                <Button type="submit">Save</Button>
            </DialogActions>
        </Dialog>)
}

const IngredientComponent = function () {
    const navigate = useNavigate();
    const [ingredients, setIngredients] = useState([]);
    const { setAllIngredients } = useContext(FoodContext);
    const [ingredientUpdating, setIngredientUpdating] = useState();
    const [openModal, setOpenModal] = useState(false);
    const [ingredientAlert, setIngredientAlert] = useState();
    const [selected, setSelected] = useState([]);
    const [deletingIngredient, setDeletingIngredient] = useState(false);
    const [addingDishRecord, setAddingDishRecord] = useState(false);

    useEffect(() => {
        refreshIngredients();
    }, [navigate]);

    const addDishRecord = (data) => {
        let dishRecordIngredientDTO = {
            dishRecord: {
                dishRecordDesc: data.dishRecordDesc,
                dishRecordTime: new Date(data.dishRecordTime).toISOString(),
            },
            ingredientIdQuantityList:
                selected.map(ingredientId => ({
                    ingredientName: data['ingredientName-' + ingredientId],
                    quantity: data['ingredientAmount-' + ingredientId]
                }))
        }

        dishRecordService.addDishRecord(dishRecordIngredientDTO).then(response => {
            setAddingDishRecord(false);
            setOpenModal(false);
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Record added successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const refreshIngredients = () => {
        ingredientService.getAllIngredients().then(response => {
            setIngredients(response.data);
            setAllIngredients(response.data);
        }).catch(error => {
            console.log(error);
        });
    }

    const searchIngredients = (searchText) => {
        ingredientService.searchIngredients(searchText).then(response => {
            setIngredients(response.data);
        }).catch(error => {
            console.log(error);
        });
    }

    const updateIngredient = (data) => {
        ingredientService.updateIngredient(data.ingredientId, data).then(response => {
            setIngredientUpdating();
            setOpenModal(false);
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredient " + data.ingredientName + " updated successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const addIngredient = (data) => {
        if (addingDishRecord) return addDishRecord(data);
        if (ingredientUpdating) return updateIngredient(data);
        ingredientService.addIngredient(data).then(response => {
            setOpenModal(false);
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredient " + data.ingredientName + " added successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const deleteIngredients = (ingredientIds) => {
        ingredientService.deleteIngredients(ingredientIds).then(response => {
            refreshIngredients();
            setIngredientAlert({ severity: "success", message: "Ingredients deleted successfully!" });
        }).catch(error => {
            console.log(error);
        });
    }

    const columns = [
        {
            field: 'ingredientId',
            headerName: 'Id',
            with: 0,
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientName',
            headerName: 'Ingredient',
            flex: 2,
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientStorage',
            headerName: 'Storage',
            flex: 2,
            type: 'number',
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientCost',
            headerName: 'Cost',
            flex: 2,
            type: 'number',
            editable: false,
            headerAlign: 'center'
        },
        {
            field: 'ingredientDesc',
            headerName: 'Description',
            flex: 3,
            editable: false,
            headerAlign: 'center'
        }
    ];

    return (
        <Box sx={{ flexGrow: 1 }}>
            <Grid container spacing={0} >
                <Grid size={12}>
                    {ingredientAlert && <Alert severity={ingredientAlert.severity} onClose={() => { setIngredientAlert(null) }}>{ingredientAlert.message}</Alert>}
                </Grid>
                <Grid size={12}>
                    <EnhancedTableToolbar numSelected={selected.length} tableTitle="Ingredients"
                        deleteSelected={() => setDeletingIngredient(true)}
                        updateSelected={() => {
                            let ingredient = ingredients.find(ingredient => ingredient.ingredientId == selected[0]);
                            setIngredientUpdating(ingredient);
                            setOpenModal(true);
                        }}
                        onSearch={(text) => { searchIngredients(text); }} />
                </Grid>
                <Grid size={12}>
                    <DataGrid
                        autoWidth
                        rows={ingredients}
                        selectionModel={selected}
                        onRowSelectionModelChange={
                            (newSelection) =>
                                newSelection.type === 'include' ? setSelected([...newSelection.ids]) :
                                    setSelected(ingredients.filter(ingredient => !newSelection.ids.has(ingredient.ingredientId)).map(ingredient => ingredient.ingredientId))
                        }
                        columns={columns}
                        initialState={{
                            sorting: {
                                sortModel: [{
                                    field: 'ingredientStorage',
                                    sort: 'asc',
                                }],
                            },
                            columns: {
                                columnVisibilityModel: {
                                    ingredientId: false,
                                },
                            },
                            pagination: {
                                paginationModel: {
                                    pageSize: 10,
                                },
                            },
                        }}
                        pageSizeOptions={[5, 10, 15, 20, 25]}
                        checkboxSelection
                        getRowId={(row) => row.ingredientId}
                        sx={{
                            boxShadow: 2,
                            border: 2,
                            borderColor: 'lightgrey',
                            width: '100%',
                            '& .MuiDataGrid-cell': {
                                textAlign: 'center', // Center the text in each cell
                            }
                        }}
                    />
                </Grid>
                {selected.length > 0 && <Button variant="contained" color="info"
                    onClick={() => {
                        setAddingDishRecord(true);
                        setOpenModal(true);
                    }}
                    sx={{
                        marginLeft: 0,
                        marginRight: 'auto',
                        marginTop: 1
                    }}
                >
                    Add Record
                </Button>}
                <Button variant="contained" color="success"
                    onClick={() => {
                        setOpenModal(true);
                    }}
                    sx={{
                        marginRight: 0,
                        marginLeft: 'auto',
                        marginTop: 1
                    }}
                >
                    New Ingredient
                </Button>
            </Grid>
            <IngredientsModal open={openModal} ingredientUpdating={ingredientUpdating} addingDishRecord={addingDishRecord} ingredientIdsForRecord={selected} ingredients={ingredients}
                onClose={() => {
                    setIngredientUpdating();
                    setAddingDishRecord(false);
                    setOpenModal(false);
                }}
                onSubmit={(event) => {
                    event.preventDefault();
                    const formData = new FormData(event.currentTarget);
                    const formJson = Object.fromEntries(formData.entries());
                    addIngredient(formJson);
                }}
            />
            <DeletionConfirmationComponent warningMessage={deletingIngredient ? "Are you sure you want to delete selected " + (selected.length > 1 ? selected.length + " ingredients?" : "ingredient?") : "Processing"}
                open={deletingIngredient} onClose={() => setDeletingIngredient(false)}
                onConfirm={() => {
                    deleteIngredients(selected);
                    setDeletingIngredient(false);
                }} />
        </Box>
    )
}

export default IngredientComponent;