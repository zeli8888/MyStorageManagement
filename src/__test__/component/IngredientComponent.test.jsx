import { render, screen, waitFor, cleanup, fireEvent } from '@testing-library/react';
import { describe, test, vi, beforeEach, expect } from 'vitest';
import IngredientComponent from '../../component/IngredientComponent';
import { FoodContext } from '../../component/FoodProvider';
import ingredientService from '../../service/IngredientService';
import '@testing-library/jest-dom/vitest';

// Mock navigation hook
vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
}));

// Mock service layer
vi.mock('../../service/IngredientService', () => ({
    default: {
        getAllIngredients: vi.fn().mockResolvedValue({ data: [] }),
        searchIngredients: vi.fn().mockResolvedValue({ data: [] }),
        addIngredient: vi.fn().mockResolvedValue({}),
        updateIngredient: vi.fn().mockResolvedValue({}),
        deleteIngredients: vi.fn().mockResolvedValue({}),
    },
}));

// Mock complex child components
vi.mock('../../component/utils', () => ({
    EnhancedTableToolbar: ({ tableTitle, onSearch, deleteSelected, updateSelected }) => (
        <div>
            {tableTitle}
            <input
                data-testid="search-input"
                onChange={(e) => onSearch(e.target.value)}
            />
            <button data-testid="delete-button" onClick={deleteSelected}>Delete</button>
            <button data-testid="update-button" onClick={updateSelected}>Update</button>
        </div>
    ),
    DeletionConfirmationComponent: ({ onConfirm, onClose }) => (
        <div>
            <button data-testid="confirm-delete" onClick={onConfirm}>Confirm</button>
            <button data-testid="cancel-delete" onClick={onClose}>Cancel</button>
        </div>
    ),
}));

// Mock DataGrid with selection handling
vi.mock('@mui/x-data-grid', () => ({
    DataGrid: ({ rows, onRowSelectionModelChange }) => (
        <div data-testid="datagrid">
            {rows.map(row => (
                <div key={row.ingredientId}>{row.ingredientName}</div>
            ))}
            <button
                data-testid="select-row"
                onClick={() => onRowSelectionModelChange({ ids: [1] })}
            >
                Select Row
            </button>
        </div>
    ),
}));

describe('IngredientComponent Integration Tests', () => {
    const mockIngredient = {
        ingredientId: 1,
        ingredientName: 'Test Ingredient',
        ingredientStorage: 5,
        ingredientCost: 10,
        ingredientDesc: 'Test Description'
    };

    beforeEach(() => {
        vi.clearAllMocks();
        ingredientService.getAllIngredients.mockResolvedValue({ data: [mockIngredient] });
    });
    afterEach(cleanup);

    // Base snapshot verification
    test('initial render matches snapshot', async () => {
        const { asFragment } = render(
            <FoodContext.Provider value={{ setAllIngredients: vi.fn() }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        await waitFor(() => expect(screen.getByText('Test Ingredient')).toBeInTheDocument());
        expect(asFragment()).toMatchSnapshot();
    });

    // Search functionality
    test('search triggers API call', async () => {
        render(
            <FoodContext.Provider value={{ setAllIngredients: vi.fn() }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        await waitFor(() => screen.getByTestId('search-input'));
        fireEvent.change(screen.getByTestId('search-input'), { target: { value: 'test' } });

        await waitFor(() => {
            expect(ingredientService.searchIngredients).toHaveBeenCalledWith('test');
        });
    });

    // Add functionality
    test('new ingredient submission', async () => {
        render(
            <FoodContext.Provider value={{ setAllIngredients: vi.fn() }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        fireEvent.click(screen.getByText(/New Ingredient/i));
        fireEvent.change(screen.getByLabelText(/Ingredient Name/i), { target: { value: 'NewItem' } });
        fireEvent.change(screen.getByLabelText(/Ingredient Storage/i), { target: { value: '5' } });
        fireEvent.click(screen.getByText(/Save/i));

        await waitFor(() => {
            expect(ingredientService.addIngredient).toHaveBeenCalledWith({
                ingredientName: 'NewItem',
                ingredientStorage: '5',
                ingredientCost: '',
                ingredientDesc: ''
            });
        });
    });

    // Update functionality
    test('existing ingredient update', async () => {
        render(
            <FoodContext.Provider value={{ setAllIngredients: vi.fn() }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        await waitFor(() => screen.getByTestId(/select-row/i));
        fireEvent.click(screen.getByTestId(/select-row/i));
        fireEvent.click(screen.getByTestId(/update-button/i));
        fireEvent.change(screen.getByLabelText(/Ingredient Name/i), { target: { value: 'Updated' } });
        fireEvent.click(screen.getByText(/Save/i));

        await waitFor(() => {
            expect(ingredientService.updateIngredient).toHaveBeenCalledWith('1', {
                ingredientId: '1',
                ingredientName: 'Updated',
                ingredientStorage: '5',
                ingredientCost: '10',
                ingredientDesc: 'Test Description'
            });
        });
    });

    // Delete functionality
    test('ingredient deletion flow', async () => {
        render(
            <FoodContext.Provider value={{ setAllIngredients: vi.fn() }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        await waitFor(() => screen.getByTestId('select-row'));
        fireEvent.click(screen.getByTestId('select-row'));
        fireEvent.click(screen.getByTestId('delete-button'));
        fireEvent.click(screen.getByTestId('confirm-delete'));

        await waitFor(() => {
            expect(ingredientService.deleteIngredients).toHaveBeenCalledWith([1]);
        });
    });

    // Error state test
    test('error alert', async () => {
        const consoleErrorMock = vi.spyOn(console, 'log').mockImplementation(() => { });
        ingredientService.getAllIngredients.mockRejectedValue(new Error('test Error'));

        let mockSetAllIngredients = vi.fn();
        render(
            <FoodContext.Provider value={{ setAllIngredients: mockSetAllIngredients }}>
                <IngredientComponent />
            </FoodContext.Provider>
        );

        await waitFor(() => {
            expect(mockSetAllIngredients).not.toHaveBeenCalled();
        });
        expect(consoleErrorMock).toHaveBeenCalledWith(new Error('test Error'));
        consoleErrorMock.mockRestore();
    });
});
