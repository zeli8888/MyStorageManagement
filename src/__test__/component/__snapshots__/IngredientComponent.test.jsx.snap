// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IngredientComponent Integration Tests > initial render matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-d64g0g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 css-yqiqnf-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 css-yqiqnf-MuiGrid-root"
      >
        <div>
          Ingredients
          <input
            data-testid="search-input"
          />
          <button
            data-testid="delete-button"
          >
            Delete
          </button>
          <button
            data-testid="update-button"
          >
            Update
          </button>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 css-yqiqnf-MuiGrid-root"
      >
        <div
          data-testid="datagrid"
        >
          <div>
            Test Ingredient
          </div>
          <button
            data-testid="select-row"
          >
            Select Row
          </button>
          <button
            data-testid="select-exclude-row"
          >
            Select Row With Exclude
          </button>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-qlt66a-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        New Ingredient
      </button>
    </div>
    <div>
      <button
        data-testid="confirm-delete"
      >
        Confirm
      </button>
      <button
        data-testid="cancel-delete"
      >
        Cancel
      </button>
    </div>
  </div>
</DocumentFragment>
`;
